### PR TITLE
BROOKLYN-278: fix Propagator validation/defaults

### PR DESCRIPTION
**Pull request by aledsage**

### CODE REVIEW

File: core/src/main/java/org/apache/brooklyn/enricher/stock/Propagator.java [hunk 1]
```diff
@@ -87,24 +93,41 @@ public Propagator() {
...
+        
+        if (propagating.size() > 0) {
+            for (Sensor<?> sensor : propagating) {
+                if (!sensorMapping.containsKey(sensor)) {
+                    sensorMapping.put(sensor, sensor);
                 }
             }
-            this.sensorMapping = ImmutableMap.copyOf(sensorMappingTemp);
-            this.sensorFilter = new Predicate<Sensor<?>>() {
+            this.sensorMapping = ImmutableMap.copyOf(sensorMapping);
```
Comment (geomacy) [inline core/src/main/java/org/apache/brooklyn/enricher/stock/Propagator.java]: I would get rid of the "this." myself - it's superfluous here and below. In particular on this line and L140 it gives the visual impression that "this.sensorMapping" is different from just "sensorMapping" which is not the case - it's the same variable in both cases.

File: core/src/main/java/org/apache/brooklyn/enricher/stock/Propagator.java [hunk 2]
```diff
@@ -113,33 +136,15 @@ public void setEntity(EntityLocal entity) {
...
-                propagatingAll = true;
-                // user specified nothing, so *set* the all_but to the default set
-                // if desired, we could allow this to be dynamically reconfigurable, remove this field and always look up;
-                // slight performance hit (always looking up), and might need to recompute subscriptions, so not supported currently
-                // TODO this default is @Beta behaviour! -- maybe better to throw?
-                propagatingAllBut = SENSORS_NOT_USUALLY_PROPAGATED;
-            } else {
-                propagatingAllBut = getConfig(PROPAGATING_ALL_BUT);
-            }
+            assert propagatingAll;
```
Comment (geomacy) [inline core/src/main/java/org/apache/brooklyn/enricher/stock/Propagator.java]: Could this not use org.apache.brooklyn.test.Asserts or com.google.common.base.Preconditions and add a message?

**Comment by geomacy** [comment]
👍 Looks good to me, a couple of minor comments but happy for this to be committed as-is.  I also pulled and tested it, and can confirm that the example given in [BROOKLYN-278|https://issues.apache.org/jira/browse/BROOKLYN-278] is rejected by this fix.

File: core/src/main/java/org/apache/brooklyn/enricher/stock/Propagator.java (hunk 1 above)
Comment (grkvlt) [inline core/src/main/java/org/apache/brooklyn/enricher/stock/Propagator.java]: @geomacy agree, normally only use `this` when there's scope ambiguity as with most ctors

**Comment by aledsage** [comment]
Thanks @geomacy - comments incorporated, and jenkins confirms build works. Merging.

**Comment by sjcorbett** [comment]
@aledsage This seems to break `EnrichersSlightlySimplerYamlTest.testWithAppEnricher`:
```
testWithAppEnricher(org.apache.brooklyn.qa.camp.EnrichersSlightlySimplerYamlTest)  Time elapsed: 1.235 sec  <<< FAILURE!
org.apache.brooklyn.util.exceptions.PropagatedRuntimeException:
	at java.util.concurrent.FutureTask.report(FutureTask.java:122)
	at java.util.concurrent.FutureTask.get(FutureTask.java:192)
	at com.google.common.util.concurrent.ForwardingFuture.get(ForwardingFuture.java:63)
	at org.apache.brooklyn.util.core.task.BasicTask.get(BasicTask.java:361)
	at org.apache.brooklyn.util.core.task.BasicTask.getUnchecked(BasicTask.java:370)
	at org.apache.brooklyn.core.objs.proxy.InternalEntityFactory.initEntityAndDescendants(InternalEntityFactory.java:361)
	at org.apache.brooklyn.core.objs.proxy.InternalEntityFactory.createEntity(InternalEntityFactory.java:188)
	at org.apache.brooklyn.core.mgmt.internal.LocalEntityManager.createEntity(LocalEntityManager.java:145)
	at org.apache.brooklyn.camp.brooklyn.AbstractYamlTest.createAndStartApplication(AbstractYamlTest.java:128)
	at org.apache.brooklyn.camp.brooklyn.AbstractYamlTest.createAndStartApplication(AbstractYamlTest.java:123)
	at org.apache.brooklyn.camp.brooklyn.AbstractYamlTest.createAndStartApplication(AbstractYamlTest.java:119)
	at org.apache.brooklyn.qa.camp.EnrichersSlightlySimplerYamlTest.testWithAppEnricher(EnrichersSlightlySimplerYamlTest.java:57)
Caused by: java.lang.IllegalStateException: Propagator enricher Propagator{running=true, entity=BasicApplicationImpl{id=B4jktQsQ}, id=FH9VxCyW} must have 'propagating' and/or 'sensorMapping', or 'propagatingAll' or 'propagatingAllBut' set
	at org.apache.brooklyn.enricher.stock.Propagator.setEntity(Propagator.java:106)
	at org.apache.brooklyn.core.entity.AbstractEntity$BasicEnricherSupport.add(AbstractEntity.java:1746)
	at org.apache.brooklyn.core.objs.proxy.InternalEntityFactory$1.run(InternalEntityFactory.java:344)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at org.apache.brooklyn.util.core.task.BasicExecutionManager$SubmissionCallable.call(BasicExecutionManager.java:519)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
``` 
Will fix.

**Comment by sjcorbett** [comment]
Oh, I guess it wasn't caught because the test is in `brooklyn-library`. The split repos make this difficult to track.

**Pull request merged**